src/fluree/db/json_ld/shacl.cljc

```diff
@@ -541,6 +561,33 @@
         #?(:clj (Pattern/compile (or valid-flags 0))
            :cljs (js/RegExp. (or valid-flags ""))))))
 
+(defn resolve-path-type
+  "Associate each property path object with its path type in order to govern path flake resolution during validation."
+  [db path-pid]
+  (go-try
+    (if-let [path-flake (->> (<? (query-range/index-range db :spot = [path-pid]))
+                             (remove #(#{const/$xsd:anyURI} (flake/p %)))
+                             (first))]
+      (let [o (flake/o path-flake)
+            p (flake/p path-flake)]
+        (cond
+          (= p const/$sh:inversePath)     [o :inverse]
+          (= p const/$sh:alternativePath) [o :alternative]
+          (= p const/$sh:zeroOrMorePath)  [o :zero-plus]
+          (= p const/$sh:oneOrMorePath)   [o :one-plus]
+          (= p const/$sh:zeroOrOnePath)   [o :zero-one]
+          :else
+          [path-pid :predicate]))
+      [path-pid :predicate])))
+
+(defn resolve-path-types
+  [db path]
+  (go-try
+    (loop [[path-pid & r] path
```

Comment:
I think all these loops would be much more clearly expressed with some combination of built-in async sequence functions and `async/reduce` (which you've essentially reimplemented here). That would require refactoring this code to use error channels because `<?` is not compatible with those built-in async functions though. I'll let you decide whether or not that change is worth it.

Reply:
In the end I couldn't get it quite working right, and the async sequence way was about twice as long, so I think I'm going to keep it as is, though I'm open to revisiting.